storage/innobase/include/trx0purge.h

```diff
@@ -149,10 +149,11 @@ class purge_sys_t
 private:
   /** number of pending stop() calls without resume() */
   Atomic_counter<uint32_t> m_paused;
-  /** number of stop_SYS() calls without resume_SYS() */
-  Atomic_counter<uint32_t> m_SYS_paused;
-  /** number of stop_FTS() calls without resume_FTS() */
-  Atomic_counter<uint32_t> m_FTS_paused;
+  /** PAUSED_SYS * number of stop_SYS() calls without resume_SYS() +
+  number of stop_FTS() calls without resume_FTS() */
+  Atomic_relaxed<uint32_t> m_FTS_paused;
+  /** The stop_SYS() multiplier in m_FTS_paused */
+  static constexpr const uint32_t PAUSED_SYS= 1U << 16;
```

Comment:
The `Atomic_relaxed` template is an `std::atomic` wrapper that uses `std::memory_order_relaxed`. But so was the `Atomic_counter` that this single atomic field is replacing.

This code has been subjected to stress testing, and the workload includes `FULLTEXT INDEX`, which mandated this logic to begin with.

To assess whether [relaxed ordering](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering]) is acceptable, we should consider the operations that read these ‘reference counts’, that is, `wait_FTS(bool)` and `must_wait_FTS()`. The read-modify-write operations themselves should be fine (but I will add the requested debug assertions to catch any wrap-around).

Outside debug assertions, the use of `purge_sys.must_wait_FTS()` is in `trx_purge_table_acquire()`, which is covered by a shared `dict_sys.latch`. We would increment the counter as part of a DDL operation, but before acquiring an exclusive `dict_sys.latch`. So, a `purge_sys_t::close_and_reopen()` loop could be triggered slightly prematurely, before the problematic DDL operation is actually executed. Decrementing the counter is not an issue; `purge_sys.resume_FTS()` or `purge_sys.resume_SYS()` would mostly be invoked while holding an exclusive `dict_sys.latch`; `ha_innobase::delete_table()` does it outside the critical section. Still, this would only cause some extra wait in the `purge_coordinator_task`, just like at the start of a DDL operation.

There are two calls to `purge_sys.wait_FTS()`: in the above mentioned `purge_sys_t::close_and_reopen()` and in `purge_sys.clone_oldest_view()`, both invoked by the `purge_coordinator_task`. There is also a `purge_sys.clone_oldest_view<true>()` call at startup when nothing can be executing any DDL that would involve `FULLTEXT INDEX`.

Based on this analysis, I think that `std::memory_order_relaxed` is completely fine here.